[LV] Use requested calling convention for vector math routines

david-arm opened the pull request:
Some vector math routines, e.g. ArmPL, specify a particular
calling convention on the routines which can help improve
performance by specifying what registers have to be preserved
across the call.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1231,6 +1231,8 @@ void VPWidenCallRecipe::execute(VPTransformState &State) {
 
   CallInst *V = State.Builder.CreateCall(Variant, Args, OpBundles);
   setFlags(V);
+  if (CallingConv::ID ID = Variant->getCallingConv())
+    V->setCallingConv(ID);
```
I'm not sure why IRBuilder doesn't do this itself but I see this idiom used all over the place so I'll not worry about it.

That said, looking at existing uses can `Function::getCallingConv()` fail?  Other instanced just do:
```
CallInst *V = State.Builder.CreateCall(Variant, Args, OpBundles);
V->setCallingConv(Variant->getCallingConv());
```

FYI: GitHub will not let me attach this to the relevant line but for information I note just above there is the assert
```
assert(Variant != nullptr && "Can't create vector function.");
```
and yet on entry to the function we've already done
```
FunctionType *VFTy = Variant->getFunctionType();
```

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/veclib-function-calls.ll:
```diff
@@ -21,28 +21,28 @@ declare float @acosf(float)
 ; SLEEF-NEON-LABEL: define void @acos_f64
 ; SLEEF-NEON-SAME: (ptr noalias [[IN_PTR:%.*]], ptr noalias [[OUT_PTR:%.*]]) #[[ATTR0:[0-9]+]] {
-; SLEEF-NEON:    [[TMP3:%.*]] = call <2 x double> @_ZGVnN2v_acos(<2 x double> [[WIDE_LOAD:%.*]])
+; SLEEF-NEON:    [[TMP2:%.*]] = call <2 x double> @_ZGVnN2v_acos(<2 x double> [[WIDE_LOAD:%.*]])
```
That's going to take quite a long time. :) It would be easier just to regenerate the test checks in a pre-commit.

david-arm replied:
https://github.com/llvm/llvm-project/pull/136138

david-arm committed:
[LV] Use requested calling convention for vector math routines

Some vector math routines, e.g. ArmPL, specify a particular
calling convention on the routines which can help improve
performance by specifying what registers have to be preserved
across the call.

fhahn reviewed (approved):
LGTM, thanks

david-arm commented:
Rebase + ran make check-all locally (which passed).
